test/sandbox.js

```diff
@@ -81,15 +86,42 @@ function run_code(code, toplevel, timeout) {
     }
 }
 
+function run_code_exec(code, toplevel, timeout) {
+    try {
+        return execSync('"' + process.argv[0] + '"', {
+            encoding: "utf8",
+            input: wrap_code(code, toplevel),
+            stdio: "pipe",
+            timeout: timeout || 5000,
+        });
+    } catch (ex) {
+        var msg = ex.message.replace(/\r\n/g, "\n");
+        var match = /\n([^:\s]*Error)(?:: ([\s\S]+?))?\n(    at [\s\S]+)\n$/.exec(msg);
+        if (match) {
+            ex = new global[match[1]](match[2]);
+            ex.stack = ex.stack.slice(0, ex.stack.indexOf("    at ")) + match[3];
+        } else if (/ETIMEDOUT/.test(msg)) {
+            ex = new Error("Script execution timed out.");
+        } else if (match = /\n *\^\n(\n.+|[^{[].*)\n/.exec(msg)) try {
+            ex = vm.runInNewContext("(" + match[1].trim() + ")");
+        } catch (e) {
+            ex = match[1];
+        }
+        return ex;
+    }
+}
+
 exports.run_code = semver.satisfies(process.version, "0.8") ? function(code, toplevel, timeout) {
-    var stdout = run_code(code, toplevel, timeout);
+    var stdout = run_code_vm(code, toplevel, timeout);
     if (typeof stdout != "string" || !/arguments/.test(code)) return stdout;
     do {
         var prev = stdout;
-        stdout = run_code(code, toplevel, timeout);
+        stdout = run_code_vm(code, toplevel, timeout);
     } while (prev !== stdout);
     return stdout;
-} : run_code;
+} : semver.satisfies(process.version, "<8") ? run_code_vm : function(code, toplevel, timeout) {
+    return (/async function/.test(code) ? run_code_exec : run_code_vm)(code, toplevel, timeout);
```

Comment:
I like the fact that it differentiates between async test cases and regular ones for perf reasons, but that regex won't catch async arrow functions or when the token span multiple lines.

Reply:
Yeah I forgot to append `[WIP]` to this PR &minus; not planning to merge this, just thought I'd clean up the stage a bit so you can have a go when you want to do so 😉